jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/Authentication.java

```diff
@@ -0,0 +1,228 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.security;
+
+import org.eclipse.jetty.io.QuietException;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+
+/**
+ * The Authentication state of a request.
+ * <p>
+ * The Authentication state can be one of several sub-types that
+ * reflects where the request is in the many different authentication
+ * cycles. Authentication might not yet be checked or it might be checked
+ * and failed, checked and deferred or succeeded.
+ */
+public interface Authentication
+{
+    static Authentication getAuthentication(Request request)
+    {
+        Object auth = request.getAttribute(Authentication.class.getName());
+        return auth instanceof Authentication authentication ? authentication : null;
+    }
+
+    static void setAuthentication(Request request, Authentication authentication)
+    {
+        request.setAttribute(Authentication.class.getName(), authentication);
+    }
+
+    class Failed extends QuietException.Exception
+    {
+        public Failed(String message)
+        {
+            super(message);
+        }
+    }
+
+    /**
+     * A successful Authentication with User information.
+     */
+    interface User extends LogoutAuthentication
```

Comment:
Some of these nested interfaces are called `Authentication.LoginAuthentication` (so they repeat "Authentication"), some don't.
We should use one style only, I think the short one is my favorite, as long as it is used elsewhere as `Authentication.User` not just `User`.

Reply:
Agree... but i may leave changing that until I've refactored eeN to use these, as they are old names and once this is used, then refactoring will do the renames.